File: web/src/lib/verify.ts
// Browser-side port of trail/verify.py for the public /verify page: unpacks a .trail.tar.gz
// (or .trail.tar.gz) and checks manifest hashes, event hashes, chain links, checkpoint heads
// and Ed25519 signatures (WebCrypto when the browser supports Ed25519; SKIP otherwise).
//
// The key inside the archive is never trusted on its own: with a TrustAnchor (Trail's
// published public key and transparency log, fetched by the page from the same origin) the
// verifier also checks that the archive was signed by Trail's key and that every checkpoint
// is in the public log. Without an anchor those checks are reported as SKIP, never PASS.
import type { VerifyCheck, VerifyResult, VerifySummary } from './types';
import { canonical, GENESIS_HASH } from './canonical';
import { sha256Bytes, sha256Hex, sha256HexBytes } from './sha256';

/** One plain-English line per check name, for readers who are not engineers. */
export const PLAIN: Record<string, string> = {
  archive: 'The file could be opened as a Trail record',
  manifest: 'The record lists its own contents',
  events: 'The events file could be read',
  'manifest file hashes': 'Every file in the pack still matches its fingerprint',
  'event hashes': 'Every event still matches its own fingerprint',
  'chain linkage': 'Every event points at the one before it, so nothing was inserted, removed or edited',
  'event count': 'The pack holds as many events as it says it does',
  "signed by Trail's key": "The signatures come from Trail's key, fetched from this site rather than from inside the file",
  checkpoints: 'Each signed checkpoint matches the chain as it stood at that moment',
  'checkpoint linkage and heads': 'Each signed checkpoint matches the chain as it stood at that moment',
  'checkpoint coverage': 'Some writing sessions had not been checkpointed yet when the pack was built',
  'checkpoint signatures': "Trail's signatures on the checkpoints are valid",
  'checkpoints in the public log': "Every checkpoint appears in Trail's public, append-only log",
  'manifest signature': 'The pack as a whole was signed by Trail',
};

class Report {
  checks: VerifyCheck[] = [];
  failed = 0;
  ok(name: string, detail = '') { this.checks.push({ status: 'PASS', name, detail, plain: PLAIN[name] }); }
  fail(name: string, detail = '') { this.failed += 1; this.checks.push({ status: 'FAIL', name, detail, plain: PLAIN[name] }); }
  skip(name: string, detail = '') { this.checks.push({ status: 'SKIP', name, detail, plain: PLAIN[name] }); }
}

// ---- gzip + tar ------------------------------------------------------------------------------------

async function gunzip(data: Uint8Array): Promise<Uint8Array> {
  if (typeof DecompressionStream === 'undefined') throw new Error('this browser cannot decompress gzip');
  const ds = new DecompressionStream('gzip');
  const stream = new Blob([data as BlobPart]).stream().pipeThrough(ds);
  return new Uint8Array(await new Response(stream).arrayBuffer());
}

function untar(buf: Uint8Array): Map<string, Uint8Array> {
  const files = new Map<string, Uint8Array>();
  const dec = new TextDecoder();
  let off = 0;
  while (off + 512 <= buf.length) {
    const header = buf.subarray(off, off + 512);
    if (header.every((b) => b === 0)) break;
    let name = dec.decode(header.subarray(0, 100)).replace(/\0.*$/s, '');
    const size = parseInt(dec.decode(header.subarray(124, 136)).replace(/\0.*$/s, '').trim() || '0', 8);
    const type = String.fromCharCode(header[156]);
    const magic = dec.decode(header.subarray(257, 263));
    if (magic.startsWith('ustar')) {
      const prefix = dec.decode(header.subarray(345, 500)).replace(/\0.*$/s, '');
      if (prefix) name = prefix + '/' + name;
    }
    off += 512;
    const body = buf.subarray(off, off + size);
    if (type === '0' || type === '\0' || type === '') files.set(name, body);
    off += Math.ceil(size / 512) * 512;
  }
  return files;
}

function readJsonl(bytes: Uint8Array): any[] {
  return new TextDecoder().decode(bytes).split('\n').map((l) => l.trim()).filter(Boolean).map((l) => JSON.parse(l));
}

// ---- Ed25519 ---------------------------------------------------------------------------------------------

type Verifier = { keyId: string; verify: (body: any, sigB64: string) => Promise<boolean> } | null;

function pemToDer(pem: string): Uint8Array {
  const b64 = pem.replace(/-----[^-]+-----/g, '').replace(/\s+/g, '');
  return Uint8Array.from(atob(b64), (c) => c.charCodeAt(0));
}

async function loadVerifier(pem: string): Promise<Verifier> {
  const der = pemToDer(pem);
  // SPKI for Ed25519 is 44 bytes: 12-byte header + 32-byte raw key.
  if (der.length !== 44) return null;
  const raw = der.subarray(12);
  const keyId = sha256HexBytes(raw).slice(0, 16);
  try {
    const key = await crypto.subtle.importKey('spki', der as BufferSource, { name: 'Ed25519' } as any, false, ['verify']);
    return {
      keyId,
      verify: async (body, sigB64) => {
        try {
          const sig = Uint8Array.from(atob(sigB64), (c) => c.charCodeAt(0));
          const digest = sha256Bytes(new TextEncoder().encode(canonical(body)));
          return await crypto.subtle.verify({ name: 'Ed25519' } as any, key, sig as BufferSource, digest as BufferSource);
        } catch {
          return false;
        }
      },
    };
  } catch {
    return { keyId, verify: null as any };
  }
}

// ---- trust anchor --------------------------------------------------------------------------------------------

export type TrustAnchor = {
  origin: string; // where the key and log came from, for the messages
  publicKeyPem: string | null; // GET /v1/public-key, or null when it could not be fetched
  transparency: { checkpoint_hash: string }[] | null; // GET /v1/transparency, or null
};

function normalisePem(pem: string): string {
  return pem.replace(/-----[^-]+-----/g, '').replace(/\s+/g, '');
}

// ---- verify ---------------------------------------------------------------------------------------------------

/** Unpack a .trail.tar.gz (or plain tar) into its files. Throws when the bytes are not an archive. */
export async function unpackRecord(data: Uint8Array): Promise<Map<string, Uint8Array>> {
  const isGzip = data[0] === 0x1f && data[1] === 0x8b;
  return untar(isGzip ? await gunzip(data) : data);
}

export async function verifyRecordBytes(data: Uint8Array, trust?: TrustAnchor): Promise<VerifyResult> {
  let files: Map<string, Uint8Array>;
  try {
    files = await unpackRecord(data);
  } catch (e) {
    const r = new Report();
    r.fail('archive', `could not read archive: ${(e as Error).message}`);
    return { ok: false, checks: r.checks };
  }
  return verifyRecordFiles(files, trust);
}

function rootOf(files: Map<string, Uint8Array>): string {
  const names = [...files.keys()];
  const roots = new Set(names.map((n) => n.split('/')[0]));
  return roots.size === 1 && names.every((n) => n.includes('/')) ? [...roots][0] + '/' : '';
}

/**
 * Alter one event in an unpacked record the way a forger would: change its text and recompute
 * its own hash so the event looks self-consistent. The next event's `prev` link no longer
 * matches, so the chain check names exactly where the record was changed. Returns the altered
 * files (a copy) and which event was touched, or null when there is nothing to alter.
 */
export function tamperRecordFiles(files: Map<string, Uint8Array>): { files: Map<string, Uint8Array>; seq: number; index: number; session: string; ts: string | null; before: string; after: string } | null {
  const root = rootOf(files);
  const dec = new TextDecoder();
  const manifestBytes = files.get(root + 'manifest.json');
  const manifest = manifestBytes ? JSON.parse(dec.decode(manifestBytes)) : {};
  const evName = root + (manifest.events_file || 'events.jsonl');
  const evBytes = files.get(evName);
  if (!evBytes) return null;
  const events: any[] = readJsonl(evBytes);
  // pick a typed insert somewhere in the middle of the record so the break is visibly mid-essay
  const candidates = events.map((e, i) => [e, i] as const).filter(([e]) => (e.kind === 'insert' || e.kind === 'paste') && typeof e.data?.text === 'string' && e.data.text.trim().length >= 3);
  if (!candidates.length) return null;
  const [ev, index] = candidates[Math.floor(candidates.length * 0.6)];
  const before = String(ev.data.text);
  // swap the first letter for a different one
  const m = before.match(/[A-Za-z]/);
  const at = m?.index ?? 0;
  const ch = before[at] || 'a';
  const swap = ch === 'e' ? 'a' : 'e';
  const after = before.slice(0, at) + swap + before.slice(at + 1);
  const altered = { ...ev, data: { ...ev.data, text: after } };
  const { hash: _drop, ...body } = altered;
  altered.hash = sha256Hex(canonical(body));
  events[index] = altered;
  const out = new Map(files);
  out.set(evName, new TextEncoder().encode(events.map((e) => JSON.stringify(e)).join('\n') + '\n'));
  return { files: out, seq: ev.seq, index, session: ev.session, ts: ev.ts ?? null, before, after };
}

export async function verifyRecordFiles(files: Map<string, Uint8Array>, trust?: TrustAnchor): Promise<VerifyResult> {
  const r = new Report();
  // Strip the single top-level directory.
  const root = rootOf(files);
  const get = (rel: string) => files.get(root + rel);
  const dec = new TextDecoder();

  const manifestBytes = get('manifest.json');
  if (!manifestBytes) {
    r.fail('manifest', 'manifest.json missing');
    return { ok: false, checks: r.checks };
  }
  const manifest = JSON.parse(dec.decode(manifestBytes));

  // 1. file hashes
  const bad: string[] = [];
  const listed = Object.keys(manifest.files || {}).sort();
  for (const rel of listed) {
    const f = get(rel);
    if (!f) { bad.push(`${rel} missing`); continue; }
    if (sha256HexBytes(f) !== manifest.files[rel]) bad.push(`${rel} hash mismatch`);
  }
  if (bad.length) r.fail('manifest file hashes', bad.slice(0, 5).join('; '));
  else r.ok('manifest file hashes', `${listed.length} files`);

  // 2 + 3. events and chains
  const evBytes = get(manifest.events_file || 'events.jsonl');
  const events: any[] = evBytes ? readJsonl(evBytes) : [];
  const bySession = new Map<string, any[]>();
  let hashErrors = 0;
  let brokenAt: VerifySummary['brokenAt'] = null;
  const mark = (ev: any, index: number) => {
    if (!brokenAt) brokenAt = { seq: ev.seq, index, session: ev.session, ts: ev.ts ?? null };
  };
  events.forEach((ev, i) => {
    const { hash, ...body } = ev;
    if (sha256Hex(canonical(body)) !== hash) {
      hashErrors += 1;
      mark(ev, i);
    }
    if (!bySession.has(ev.session)) bySession.set(ev.session, []);
    bySession.get(ev.session)!.push(ev);
  });
  if (hashErrors) r.fail('event hashes', `${hashErrors} of ${events.length} events do not match their contents; first at event #${brokenAt!.index + 1}`);
  else r.ok('event hashes', `${events.length} events`);

  const chainErrors: string[] = [];
  const sessionOrder = [...bySession.keys()];
  for (const [session, evs] of [...bySession.entries()].sort((a, b) => (a[0] < b[0] ? -1 : 1))) {
    evs.sort((a, b) => a.seq - b.seq);
    let prev = GENESIS_HASH;
    evs.forEach((ev, i) => {
      const index = events.indexOf(ev);
      if (ev.seq !== i) { chainErrors.push(`session ${sessionOrder.indexOf(session) + 1}: seq gap at ${i}`); mark(ev, index); }
      if (ev.prev !== prev) { chainErrors.push(`event #${index + 1} (session ${sessionOrder.indexOf(session) + 1}, seq ${ev.seq}) points at a previous event that no longer exists: the record was changed at or before event #${index}`); mark(events[index - 1] ?? ev, Math.max(0, index - 1)); }
      prev = ev.hash;
    });
  }
  if (chainErrors.length) r.fail('chain linkage', chainErrors.slice(0, 3).join('; '));
  else r.ok('chain linkage', `${bySession.size} session${bySession.size === 1 ? '' : 's'}, ${events.length} links`);
  if (manifest.event_count != null && manifest.event_count !== events.length) {
    r.fail('event count', `manifest says ${manifest.event_count}, archive has ${events.length}`);
  }

  // 4. checkpoints
  const cpBytes = get(manifest.checkpoints_file || 'checkpoints.jsonl');
  const checkpoints: any[] = cpBytes ? readJsonl(cpBytes) : [];
  const pkBytes = get(manifest.public_key_file || 'keys/signer.pub.pem');
  const archivePem = pkBytes ? dec.decode(pkBytes) : null;
  const verifier = archivePem ? await loadVerifier(archivePem) : null;
  const canSign = !!(verifier && verifier.verify);
  const sigBytes = get('manifest.sig.json');

  // Is the key in the archive Trail's key? Anyone can generate a key and sign a fabricated
  // record with it, so this check decides what "Verified" means.
  if (trust) {
    if (!checkpoints.length && !sigBytes) {
      r.fail("signed by Trail's key", 'nothing in this archive carries a signature');
    } else if (!archivePem || !verifier) {
      r.fail("signed by Trail's key", 'the archive carries no usable public key');
    } else if (trust.publicKeyPem === null) {
      r.skip("signed by Trail's key", `could not fetch Trail's public key from ${trust.origin}; compare key id ${verifier.keyId} with ${trust.origin}/v1/public-key yourself`);
    } else {
      const trusted = await loadVerifier(trust.publicKeyPem);
      if (trusted && trusted.keyId === verifier.keyId && normalisePem(trust.publicKeyPem) === normalisePem(archivePem)) {
        r.ok("signed by Trail's key", `key id ${verifier.keyId} matches ${trust.origin}/v1/public-key`);
      } else {
        r.fail("signed by Trail's key", `the archive is signed by an unknown key (${verifier.keyId}); Trail's key is ${trusted?.keyId ?? 'unavailable'}. Hashes and chains below may still be internally consistent, but this pack could have been made by anyone`);
      }
    }
  }

  if (!checkpoints.length) {
    r.skip('checkpoints', 'none in archive');
  } else {
    const cpErrors: string[] = [];
    checkpoints.sort((a, b) => a.body.seq - b.body.seq);
    const firstSeq = checkpoints[0].body.seq;
    let prevHash: string | null = firstSeq === 0 ? GENESIS_HASH : null;
    const covered = new Set<string>();
    for (let i = 0; i < checkpoints.length; i++) {
      const cp = checkpoints[i];
      const body = cp.body;
      const bodyHash = sha256Hex(canonical(body));
      if ((cp.hash ?? bodyHash) !== bodyHash) cpErrors.push(`checkpoint ${body.seq}: hash mismatch`);
      if (body.seq !== firstSeq + i) cpErrors.push(`checkpoint ${body.seq}: expected seq ${firstSeq + i}`);
      if (prevHash !== null && body.prev_checkpoint !== prevHash) cpErrors.push(`checkpoint ${body.seq}: prev_checkpoint mismatch`);
      prevHash = bodyHash;
      for (const [session, head] of Object.entries<any>(body.heads || {})) {
        const evs = bySession.get(session);
        if (!evs) { cpErrors.push(`checkpoint ${body.seq}: session ${session} is named but missing from the record`); continue; }
        covered.add(session);
        if (head.seq >= evs.length || evs[head.seq].hash !== head.hash) cpErrors.push(`checkpoint ${body.seq}: head of ${session} does not match archive`);
        else if (head.first_ts != null && evs[0].ts !== head.first_ts) cpErrors.push(`checkpoint ${body.seq}: first_ts of ${session} does not match archive`);
      }
      if (canSign) {
        if (cp.key_id !== verifier!.keyId) cpErrors.push(`checkpoint ${body.seq}: signed by a different key (${cp.key_id})`);
        else if (!(await verifier!.verify(body, cp.signature))) cpErrors.push(`checkpoint ${body.seq}: bad signature`);
      }
    }
    if (cpErrors.length) r.fail('checkpoints', cpErrors.slice(0, 5).join('; '));
    else r.ok('checkpoint linkage and heads', `${checkpoints.length} checkpoints, ${covered.size} of ${bySession.size} sessions covered`);
    const uncovered = bySession.size - covered.size;
    if (uncovered) r.skip('checkpoint coverage', `${uncovered} session(s) not yet named by any checkpoint in this archive`);
    if (!canSign) r.skip('checkpoint signatures', 'this browser cannot verify Ed25519 signatures; run verify.py to check them');
    else r.ok('checkpoint signatures', `key ${verifier!.keyId}`);

    // public log: every checkpoint we signed is appended to /v1/transparency
    if (trust) {
      if (trust.transparency === null) {
        r.skip('checkpoints in the public log', `could not fetch ${trust.origin}/v1/transparency`);
      } else {
        const logged = new Set(trust.transparency.map((e) => e.checkpoint_hash));
        const missing = checkpoints.filter((cp) => !logged.has(cp.hash ?? sha256Hex(canonical(cp.body))));
        if (missing.length) r.fail('checkpoints in the public log', `${missing.length} of ${checkpoints.length} checkpoints are not in Trail's transparency log`);
        else r.ok('checkpoints in the public log', `all ${checkpoints.length} found in ${trust.origin}/v1/transparency`);
      }
    }
  }

  // 5. manifest signature
  if (sigBytes) {
    const sig = JSON.parse(dec.decode(sigBytes));
    if (!canSign) r.skip('manifest signature', 'no public key or Ed25519 unavailable in this browser');
    else if (sig.key_id !== verifier!.keyId) r.fail('manifest signature', 'signed by a different key');
    else if (sig.manifest_hash !== sha256Hex(canonical(manifest))) r.fail('manifest signature', 'manifest hash in signature does not match manifest.json');
    else if (!(await verifier!.verify({ manifest_hash: sig.manifest_hash }, sig.signature))) r.fail('manifest signature', 'bad signature');
    else r.ok('manifest signature', `key ${verifier!.keyId}`);
  } else {
    r.skip('manifest signature', 'manifest.sig.json not present');
  }
  const summary: VerifySummary = {
    events: events.length,
    sessions: bySession.size,
    checkpoints: checkpoints.length,
    signedAt: checkpoints.map((cp) => String(cp.body?.ts ?? '')).filter(Boolean),
    brokenAt,
  };
  return { ok: r.failed === 0, checks: r.checks, summary };
}

/** A bare events.jsonl (raw export): event hashes and chain links only; no checkpoints to check. */
export function verifyEventsJsonl(bytes: Uint8Array): VerifyResult {
  const r = new Report();
  let events: any[];
  try {
    events = readJsonl(bytes);
  } catch (e) {
    r.fail('events', `could not parse events.jsonl: ${(e as Error).message}`);
    return { ok: false, checks: r.checks };
  }
  if (!events.length) {
    r.fail('events', 'the file has no events');
    return { ok: false, checks: r.checks };
  }
  const bySession = new Map<string, any[]>();
  let hashErrors = 0;
  for (const ev of events) {
    const { hash, ...body } = ev;
    if (sha256Hex(canonical(body)) !== hash) hashErrors += 1;
    if (!bySession.has(ev.session)) bySession.set(ev.session, []);
    bySession.get(ev.session)!.push(ev);
  }
  if (hashErrors) r.fail('event hashes', `${hashErrors} of ${events.length} events do not match their contents`);
  else r.ok('event hashes', `${events.length} events`);
  const chainErrors: string[] = [];
  for (const [session, evs] of bySession) {
    evs.sort((a, b) => a.seq - b.seq);
    let prev = GENESIS_HASH;
    evs.forEach((ev, i) => {
      if (ev.seq !== i) chainErrors.push(`session ${session}: seq gap at ${i}`);
      if (ev.prev !== prev) chainErrors.push(`session ${session} seq ${ev.seq}: prev mismatch`);
      prev = ev.hash;
    });
  }
  if (chainErrors.length) r.fail('chain linkage', chainErrors.slice(0, 5).join('; '));
  else r.ok('chain linkage', `${bySession.size} sessions`);
  r.skip('checkpoints', 'a raw events file carries no signed checkpoints; export a record (.tar.gz) to check signatures');
  return { ok: r.failed === 0, checks: r.checks };
}

export async function verifyRecordFile(file: File, trust?: TrustAnchor): Promise<VerifyResult> {
  const bytes = new Uint8Array(await file.arrayBuffer());
  const isGzip = bytes[0] === 0x1f && bytes[1] === 0x8b;
  if (!isGzip && (file.name.endsWith('.jsonl') || file.name.endsWith('.json') || bytes[0] === 0x7b)) return verifyEventsJsonl(bytes);
  return verifyRecordBytes(bytes, trust);
}

File: trail/verify.py
#!/usr/bin/env python3
"""Standalone verifier for a Trail writing record.

This file is copied verbatim into every archive so that a third party can
verify the record with nothing but this script, the archive, and the
signer's public key. It depends only on the Python standard library;
signature checks additionally need the `cryptography` package and are
reported as SKIPPED when it is absent.

Usage:
    python verify.py <archive.tar.gz | extracted directory> [--public-key signer.pub.pem]

Checks performed:
    1. Every file listed in manifest.json has the recorded SHA-256.
    2. Every event's hash equals SHA-256(canonical JSON of its body).
    3. Within each session, seq is contiguous from 0 and every event's
       `prev` equals the previous event's hash (genesis for seq 0), and
       every session started inside the archive's stated period.
    4. Every checkpoint's signature verifies under the public key, the
       checkpoints chain to one another, every session a checkpoint names
       as starting in the period is present, and every named head matches
       the event at that seq in the archive.
    5. If manifest.sig.json is present, the manifest signature verifies.

Exit code 0 means every check passed; 1 means at least one failed.
"""

from __future__ import annotations

import base64
import hashlib
import json
import os
import sys
import tarfile
import tempfile

GENESIS = "0" * 64


def canonical(value) -> bytes:
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False, allow_nan=False).encode(
        "utf-8"
    )


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def read_jsonl(path):
    out = []
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line:
                out.append(json.loads(line))
    return out


def load_verifier(pem_path):
    try:
        from cryptography.hazmat.primitives import serialization
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
        from cryptography.exceptions import InvalidSignature
    except ImportError:
        return None
    with open(pem_path, "rb") as f:
        key = serialization.load_pem_public_key(f.read())
    if not isinstance(key, Ed25519PublicKey):
        raise SystemExit("public key is not Ed25519")
    raw = key.public_bytes(serialization.Encoding.Raw, serialization.PublicFormat.Raw)
    key_id = sha256_hex(raw)[:16]

    def verify(body, sig_b64) -> bool:
        try:
            key.verify(base64.b64decode(sig_b64), hashlib.sha256(canonical(body)).digest())
            return True
        except (InvalidSignature, ValueError):
            return False

    return key_id, verify


class Report:
    def __init__(self):
        self.checks = []
        self.failed = 0

    def ok(self, name, detail=""):
        self.checks.append(("PASS", name, detail))

    def fail(self, name, detail=""):
        self.failed += 1
        self.checks.append(("FAIL", name, detail))

    def skip(self, name, detail=""):
        self.checks.append(("SKIP", name, detail))

    def print(self):
        for status, name, detail in self.checks:
            line = f"[{status}] {name}"
            if detail:
                line += f": {detail}"
            print(line)
        print("RESULT:", "VERIFIED" if self.failed == 0 else f"FAILED ({self.failed} check(s))")


def verify_directory(root: str, public_key: str | None = None) -> Report:
    r = Report()
    manifest_path = os.path.join(root, "manifest.json")
    if not os.path.exists(manifest_path):
        r.fail("manifest", "manifest.json missing")
        return r
    with open(manifest_path, "r", encoding="utf-8") as f:
        manifest = json.load(f)

    # 1. file hashes
    bad = []
    for rel, expected in sorted(manifest.get("files", {}).items()):
        p = os.path.join(root, rel)
        if not os.path.exists(p):
            bad.append(f"{rel} missing")
            continue
        with open(p, "rb") as f:
            actual = sha256_hex(f.read())
        if actual != expected:
            bad.append(f"{rel} hash mismatch")
    if bad:
        r.fail("manifest file hashes", "; ".join(bad[:5]))
    else:
        r.ok("manifest file hashes", f"{len(manifest.get('files', {}))} files")

    # 2 + 3. events and chains
    events = read_jsonl(os.path.join(root, manifest.get("events_file", "events.jsonl")))
    by_session = {}
    hash_errors = 0
    for ev in events:
        body = {k: v for k, v in ev.items() if k != "hash"}
        if sha256_hex(canonical(body)) != ev.get("hash"):
            hash_errors += 1
        by_session.setdefault(ev["session"], []).append(ev)
    if hash_errors:
        r.fail("event hashes", f"{hash_errors} of {len(events)} events do not match their contents")
    else:
        r.ok("event hashes", f"{len(events)} events")

    chain_errors = []
    heads = {}
    for session, evs in sorted(by_session.items()):
        evs.sort(key=lambda e: e["seq"])
        prev = GENESIS
        for i, ev in enumerate(evs):
            if ev["seq"] != i:
                chain_errors.append(f"session {session}: seq gap at {i}")
            if ev["prev"] != prev:
                chain_errors.append(f"session {session} seq {ev['seq']}: prev mismatch")
            prev = ev["hash"]
        heads[session] = {"seq": evs[-1]["seq"], "hash": evs[-1]["hash"]}
    if chain_errors:
        r.fail("chain linkage", "; ".join(chain_errors[:5]))
    else:
        r.ok("chain linkage", f"{len(by_session)} sessions")

    if manifest.get("event_count") is not None and manifest["event_count"] != len(events):
        r.fail("event count", f"manifest says {manifest['event_count']}, archive has {len(events)}")

    p_start = p_end = None

    # 4. checkpoints
    cp_path = os.path.join(root, manifest.get("checkpoints_file", "checkpoints.jsonl"))
    checkpoints = read_jsonl(cp_path) if os.path.exists(cp_path) else []
    pk = public_key or os.path.join(root, manifest.get("public_key_file", "keys/signer.pub.pem"))
    verifier = load_verifier(pk) if os.path.exists(pk) else None
    if not checkpoints:
        r.skip("checkpoints", "none in archive")
    else:
        key_id, verify = verifier if verifier else (None, None)
        cp_errors = []
        checkpoints.sort(key=lambda c: c["body"]["seq"])
        first_seq = checkpoints[0]["body"]["seq"]
        prev_hash = GENESIS if first_seq == 0 else None  # linkage to earlier checkpoints is outside this archive
        covered = set()
        outside_sessions = 0
        for i, cp in enumerate(checkpoints):
            body = cp["body"]
            body_hash = sha256_hex(canonical(body))
            if cp.get("hash", body_hash) != body_hash:
                cp_errors.append(f"checkpoint {body['seq']}: hash mismatch")
            if body["seq"] != first_seq + i:
                cp_errors.append(f"checkpoint {body['seq']}: expected seq {first_seq + i}")
            if prev_hash is not None and body["prev_checkpoint"] != prev_hash:
                cp_errors.append(f"checkpoint {body['seq']}: prev_checkpoint mismatch")
            prev_hash = body_hash
            for session, head in body.get("heads", {}).items():
                evs = by_session.get(session)
                first_ts = head.get("first_ts")
                in_period = bool(p_start and p_end and first_ts and p_start <= first_ts < p_end)
                if evs is None:
                    cp_errors.append(f"checkpoint {body['seq']}: session {session} is named but missing from the record")
                    continue
                covered.add(session)
                if head["seq"] >= len(evs) or evs[head["seq"]]["hash"] != head["hash"]:
                    cp_errors.append(f"checkpoint {body['seq']}: head of {session} does not match archive")
                elif first_ts is not None and evs[0]["ts"] != first_ts:
                    cp_errors.append(f"checkpoint {body['seq']}: first_ts of {session} does not match archive")
            if verify is not None:
                if cp.get("key_id") != key_id:
                    cp_errors.append(f"checkpoint {body['seq']}: signed by a different key ({cp.get('key_id')})")
                elif not verify(body, cp["signature"]):
                    cp_errors.append(f"checkpoint {body['seq']}: bad signature")
        if cp_errors:
            r.fail("checkpoints", "; ".join(cp_errors[:5]))
        else:
            r.ok("checkpoint linkage and heads", f"{len(checkpoints)} checkpoints, {len(covered)} of {len(by_session)} sessions covered")
        uncovered = len(by_session) - len(covered)
        if uncovered:
            r.skip("checkpoint coverage", f"{uncovered} session(s) not yet named by any checkpoint in this archive")
        if verify is None:
            r.skip("checkpoint signatures", "install `cryptography` to verify Ed25519 signatures")
        else:
            r.ok("checkpoint signatures", f"key {key_id}")

    # 5. manifest signature
    sig_path = os.path.join(root, "manifest.sig.json")
    if os.path.exists(sig_path):
        with open(sig_path, "r", encoding="utf-8") as f:
            sig = json.load(f)
        if verifier is None:
            r.skip("manifest signature", "no public key or cryptography unavailable")
        else:
            key_id, verify = verifier
            if sig.get("key_id") != key_id:
                r.fail("manifest signature", "signed by a different key")
            elif sig.get("manifest_hash") != sha256_hex(canonical(manifest)):
                r.fail("manifest signature", "manifest hash in signature does not match manifest.json")
            elif not verify({"manifest_hash": sig["manifest_hash"]}, sig["signature"]):
                r.fail("manifest signature", "bad signature")
            else:
                r.ok("manifest signature", f"key {key_id}")
    else:
        r.skip("manifest signature", "manifest.sig.json not present")

    return r


def verify_path(path: str, public_key: str | None = None) -> Report:
    if os.path.isdir(path):
        return verify_directory(path, public_key)
    with tempfile.TemporaryDirectory() as tmp:
        with tarfile.open(path, "r:*") as tar:
            for member in tar.getmembers():
                target = os.path.realpath(os.path.join(tmp, member.name))
                if not target.startswith(os.path.realpath(tmp) + os.sep) and target != os.path.realpath(tmp):
                    raise SystemExit(f"unsafe path in archive: {member.name}")
            tar.extractall(tmp)
        entries = os.listdir(tmp)
        root = os.path.join(tmp, entries[0]) if len(entries) == 1 and os.path.isdir(os.path.join(tmp, entries[0])) else tmp
        return verify_directory(root, public_key)


def main(argv=None) -> int:
    argv = list(sys.argv[1:] if argv is None else argv)
    if not argv or argv[0] in ("-h", "--help"):
        print(__doc__)
        return 2
    path = argv[0]
    pk = None
    if "--public-key" in argv:
        pk = argv[argv.index("--public-key") + 1]
    report = verify_path(path, pk)
    report.print()
    return 0 if report.failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())
